R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,267 @@
+    for (i in 2:length(calls)) {
+      if (grepl("@linksto", calls[[i]][1, "text"])) {
+        calls[[i - 1]] <- rbind(calls[[i - 1]], calls[[i]][1, ])
+        calls[[i]] <- calls[[i]][-1, ]
+      }
+    }
+  }
+  calls
+}
+
+# code_graph ----
+
+#' Create Object Dependencies Graph Within Parsed Code
+#'
+#' @description This function constructs a dependency graph that identifies the relationships between objects in
+#' parsed code. It helps you understand which objects are needed to recreate a specific object.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#'
+#' @return A `list` (of length of input `calls_pd`) where each element represents one call. Each element consists of a
+#' character vector containing names of objects that were influenced by this call, and names of objects influencing this
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`. If an object was marked with `@linksto` side effects tag
+#' then it appears as an influenced object at the beginning of a vector.
+#'
+#' @keywords internal
+#' @noRd
+code_graph <- function(calls_pd) {
+  cooccurence <- extract_occurrence(calls_pd)
+
+  side_effects <- extract_side_effects(calls_pd)
+
+  graph <- mapply(c, side_effects, cooccurence, SIMPLIFY = FALSE)
+
+  remove_graph_duplicates(graph)
+}
+
+#' Extract Object Occurrence
+#'
+#' @description Extract objects occurrence within calls passed by `calls_pd`. It also detects which objects are
+#' influenced by others within a call, and which are influencers.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#' @return A `list` (of length of input `calls_pd`) where each element represents one call. Each element consists of a
+#' character vector containing names of objects that were influenced by this call, and names of objects influencing this
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`.
+#' @keywords internal
+#' @noRd
+extract_occurrence <- function(calls_pd) {
+  is_in_function <- function(x) {
+    # If an object is a function parameter,
+    # then in calls_pd there is a `SYMBOL_FORMALS` entry for that object.
+    function_id <- x[x$token == "FUNCTION", "parent"]
+    if (length(function_id)) {
+      x$id %in% get_children(x, function_id)$id
+    } else {
+      rep(FALSE, nrow(x))
+    }
+  }
+  lapply(
+    calls_pd,
+    function(call_pd) {
+
+      # What occurs in a function body is not tracked.
+      x <- call_pd[!is_in_function(call_pd), ]
+      sym_cond <- which(x$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"))
```

Comment:
`which` returns `integer(0)` if `x` is all `FALSE`. Please make sure this will not kick us in the backside somewhere.

Reply:
yeah, if `sym_cond` is integer(0)  which means there were no symbols in the call then it's ok. Check this test f472e68e798008aac4a8f62a137f4fc41221230d

Reply:
also added a return statement if sym_cond is empty to fail earlier 3b53e80f044c83c0f9590c5ac6abe6b20e26ba81